internal/dag/accessors.go

```diff
@@ -0,0 +1,241 @@
+// Copyright Project Contour Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package dag
+
+import (
+	"fmt"
+	"strconv"
+
+	"github.com/projectcontour/contour/internal/annotation"
+	"github.com/projectcontour/contour/internal/k8s"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/intstr"
+)
+
+// RouteServiceName identifies a service used in a route.
+type RouteServiceName struct {
+	Name      string
+	Namespace string
+	Port      int32
+}
+
+// GetServices returns all services in the DAG.
+func GetServices(dag *DAG) map[RouteServiceName]*Service {
```

Comment:
Could this be a method off of the `dag` type so that you wouldn't need to pass in the dag each time?

Reply:
Definitely could, yes. I guess the one benefit of keeping them as standalone functions is that they could move into a different package along with the processors. But, in the current state, they could definitely be methods. Happy to change them if you think that makes sense.